### PR TITLE
SyliusReportBundle

### PR DESCRIPTION
**Pull request by Zales0123**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | 
| License       | MIT
| Doc PR        | https://github.com/Sylius/Sylius-Docs/pull/251

SyliusReportBundle - flexible reports system, based on data fetchers and renderers. Now it's extremely easy to create a report with whatever data you want and nice look. Some default data fetchers and renderers already implemented, but adding another ones is a piece of cake. Find out more: https://github.com/Sylius/Sylius-Docs/pull/251
syliusreportbundle1
syliusreportbundle2
syliusreportbundle3
syliusreportbundle4

### CODE REVIEW

**Comment by stloyd** [comment]
Guys, looks interesting, but seriously you should squash those commits few tiems, over 90 is way to much to easily review ;)

File: src/Sylius/Bundle/CoreBundle/DataFetcher/NumberOfOrdersDataFetcher.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Sylius\Bundle\CoreBundle\DataFetcher;
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/DataFetcher/NumberOfOrdersDataFetcher.php]: Missing license.

File: src/Sylius/Bundle/CoreBundle/DataFetcher/SalesTotalDataFetcher.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Sylius\Bundle\CoreBundle\DataFetcher;
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/DataFetcher/SalesTotalDataFetcher.php]: Missing license.

File: src/Sylius/Bundle/CoreBundle/DataFetcher/SalesTotalDataFetcher.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+use Sylius\Bundle\CoreBundle\Doctrine\ORM\OrderRepository;
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/DataFetcher/SalesTotalDataFetcher.php]: Why using repository & not interface?

File: src/Sylius/Bundle/ReportBundle/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+CHANGELOG
+=========
```
Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/CHANGELOG.md]: Missing initial note.

File: src/Sylius/Bundle/ReportBundle/DataFetcher/TimePeriod.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                break;
+            case self::PERIOD_MONTH:
+                $this->setExtraConfiguration($configuration, 'P1M', '%m', 'F Y', array('month', 'year'));
+                break;
+            case self::PERIOD_YEAR:
+                $this->setExtraConfiguration($configuration, 'P1Y', '%y', 'Y', array('year'));
+                break;
+            default:
+                throw new \InvalidArgumentException('Wrong data fetcher period');
+                break;
```
Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/DataFetcher/TimePeriod.php]: This is useless.

File: src/Sylius/Bundle/ReportBundle/Form/EventListener/BuildReportRendererFormListener.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    public function addConfigurationFields(FormInterface $form, $rendererType, array $data = array())
+    {
+        $renderer = $this->rendererRegistry->get($rendererType);
+        $formType = sprintf('sylius_renderer_%s', $renderer->getType());
+
+        try {
+            $configurationField = $this->factory->createNamed('rendererConfiguration', $formType, $data, array('auto_initialize' => false));
+        } catch (\InvalidArgumentException $e) {
+            var_dump($e);
```
Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/Form/EventListener/BuildReportRendererFormListener.php]: Really?

File: src/Sylius/Bundle/ReportBundle/Form/Type/ReportType.php [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            ->add('dataFetcher', 'sylius_data_fetcher_choice', array(
+                'label'    => 'sylius.form.report.data_fetcher',
+            ))
+            ->add('renderer', 'sylius_renderer_choice', array(
+                'label' => 'sylius.form.report.renderer.label'
+            ))
+            ->addEventSubscriber(new BuildReportRendererFormListener($this->rendererRegistry, $builder->getFormFactory()))
+        ;
+
+        $prototypes = array();
```
Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/Form/Type/ReportType.php]:
```php
$prototypes = array(
    'renderers' => array(),
    'dataFetchers' => array(),
);

File: src/Sylius/Bundle/ReportBundle/composer.json [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            "name":     "Community contributions",
+            "homepage": "http://github.com/Sylius/Sylius/contributors"
+        }
+    ],
+    "require": {
+        "php":                             ">=5.3.3",
+
+        "symfony/framework-bundle":        "~2.3",
+        "stof/doctrine-extensions-bundle": "1.1.*",
```
Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/composer.json]: `~1.1`

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php [hunk 9]
```diff
@@ -327,6 +327,7 @@ public function getForm($resource = null)
         if (strpos($type, '\\') !== false) { // full class name specified
             $type = new $type();
         } elseif (!$this->get('form.registry')->hasType($type)) { // form alias is not registered
+
```
Comment (stloyd) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: Should be reverted.

File: src/Sylius/Bundle/WebBundle/Menu/BackendMenuBuilder.php [hunk 10]
```diff
@@ -240,6 +240,10 @@ protected function addSalesMenu(ItemInterface $menu, array $childOptions, $secti
             'route' => 'sylius_backend_promotion_create',
             'labelAttributes' => array('icon' => 'glyphicon glyphicon-plus-sign'),
         ))->setLabel($this->translate(sprintf('sylius.backend.menu.%s.new_promotion', $section)));
+        $child->addChild('report', array(
```
Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Menu/BackendMenuBuilder.php]: Missing new line before.

File: src/Sylius/Component/Report/DataFetcher/Data.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
+<?php
+/*
```
Comment (stloyd) [inline src/Sylius/Component/Report/DataFetcher/Data.php]: Format of this and author note is wrong. Also new line before missing.

File: src/Sylius/Component/Report/DataFetcher/DataFetcherInterface.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
+<?php
+/*
+* This file is part of the Sylius package.
```
Comment (stloyd) [inline src/Sylius/Component/Report/DataFetcher/DataFetcherInterface.php]: Same.

File: src/Sylius/Component/Report/DataFetcher/DelegatingDataFetcher.php [hunk 13]
```diff
@@ -0,0 +1,54 @@
+<?php
+/*
+* This file is part of the Sylius package.
```
Comment (stloyd) [inline src/Sylius/Component/Report/DataFetcher/DelegatingDataFetcher.php]: Same.

File: src/Sylius/Component/Report/Model/ReportInterface.php [hunk 14]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace Sylius\Component\Report\Model;
```
Comment (stloyd) [inline src/Sylius/Component/Report/Model/ReportInterface.php]: Missing license.

File: src/Sylius/Component/Report/Model/ReportInterface.php [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+interface ReportInterface
+{
+    public function getId();
+    public function setId($id);
```
Comment (stloyd) [inline src/Sylius/Component/Report/Model/ReportInterface.php]: This is wrong.

File: src/Sylius/Component/Report/composer.json [hunk 16]
```diff
@@ -0,0 +1,45 @@
+{
+    "name": "sylius/product",
```
Comment (stloyd) [inline src/Sylius/Component/Report/composer.json]: Not really.

File: src/Sylius/Component/Report/composer.json [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    "type": "library",
+    "description": "Product catalog system with support for product options and variants.",
```
Comment (stloyd) [inline src/Sylius/Component/Report/composer.json]: Copy&paste in whole file.

**Comment by a2xchip** [comment]
nice bundle

File: src/Sylius/Bundle/ReportBundle/DataFetcher/TimePeriod.php [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        $configuration['groupBy'] = $groupBy;
+    }
+
+    private function fillEmptyRecodrs(array $fetched, array $configuration)
+    {
+        $date = $configuration['start'];
+        $dateInterval = new \DateInterval($configuration['interval']);
+        $numberOfPeriods = $configuration['start']->diff($configuration['end']);
+
+        for ($i = 0; $i <= $numberOfPeriods->format($configuration['periodFormat']); $i++) {
```
Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/DataFetcher/TimePeriod.php]: You should avoid function call in loop, assign it to variable:
```php
$numberOfPeriods = $configuration['start']->diff($configuration['end'])->format($configuration['periodFormat']);
for ($i = 0; $i <= $numberOfPeriods; $i++) {
}

File: src/Sylius/Component/Report/Model/Report.php [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+
+class Report implements ReportInterface
+{
+    /**
+     *@var integer
+     */
+    private $id;
+
+    /**
+     * @var String
```
Comment (stloyd) [inline src/Sylius/Component/Report/Model/Report.php]: Don't use Uppercase...

File: src/Sylius/Component/Report/Model/Report.php [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    /**
+     * Sets the value of id.
+     *
+     * @param integer $id the id
+     *
+     * @return self
+     */
+    public function setId($id)
```
Comment (stloyd) [inline src/Sylius/Component/Report/Model/Report.php]: Identifier is internal value, you should never change it directly.

File: src/Sylius/Component/Report/Model/Report.php [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+    {
+        $this->code = $code;
+
+        return $this;
+    }
+
+    /**
+     * Gets the dataFetcher name.
+     *
+     * @return DataFetcher
```
Comment (stloyd) [inline src/Sylius/Component/Report/Model/Report.php]: It's DataFetcher or string? You use both, one here, second on variable declaration.

File: src/Sylius/Bundle/ReportBundle/Controller/ReportController.php [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ReportBundle\Controller;
+
+use Sylius\Bundle\ResourceBundle\Controller\ResourceController;
+use Symfony\Component\HttpFoundation\Request;
+use Sylius\Component\Report\Model\ReportInterface;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
```
Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/Controller/ReportController.php]: Not used anywhere.

File: src/Sylius/Bundle/ReportBundle/Controller/ReportController.php [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+class ReportController extends ResourceController
+{
+    /**
+     * @param Request $request
+     *
+     * @return Response
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ReportBundle/Controller/ReportController.php]: Should be ``` Symfony\Component\HttpFoundation\Response ``` - add **use** statement

File: src/Sylius/Bundle/ReportBundle/Controller/ReportController.php [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+            $configurationForm->submit($request);
+        }
+
+        return $this->render($this->config->getTemplate('show.html'), array('report' => $report, 'form' => $configurationForm->createView(), 'configuration' => $configurationForm->getData()));
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return Response
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ReportBundle/Controller/ReportController.php]: Should be ``` Symfony\Component\HttpFoundation\Response ``` - add **use** statement

File: src/Sylius/Component/Report/Model/Report.php [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+     * Sets the value of dataFetcher.
+     *
+     * @param string $dataFetcher the data fetcher
+     *
+     * @return self
+     */
+    public function setDataFetcher($dataFetcher)
+    {
+        $this->dataFetcher = $dataFetcher;
+    }
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Report/Model/Report.php]: ``` return $this ```

File: src/Sylius/Bundle/ReportBundle/Form/Type/ReportType.php [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+                'label'    => 'sylius.form.report.description',
+                'required' => false,
+            ))
+            ->add('dataFetcher', 'sylius_data_fetcher_choice', array(
+                'label'    => 'sylius.form.report.data_fetcher',
+            ))
+            ->add('renderer', 'sylius_renderer_choice', array(
+                'label' => 'sylius.form.report.renderer.label'
+            ))
+            ->addEventSubscriber(new BuildReportRendererFormListener($this->rendererRegistry, $builder->getFormFactory()))
```
Comment (BartoszSiejka) [inline src/Sylius/Bundle/ReportBundle/Form/Type/ReportType.php]: Doubled listener

**Comment by pjedrzejewski** [comment]
@Zales0123 You use PHP 5.4 only syntax (array access) in TableRenderer and spec fails on 5.3. (see travis) Please remove these and it will turn green.

**Comment by peteward** [comment]
Just a bit of visual feedback - do the Time Period dropdown and the Filter button really need to be that loooooooooooooong?

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/SyliusCoreExtension.php [hunk 27]
```diff
@@ -59,6 +60,13 @@ class SyliusCoreExtension extends AbstractResourceExtension implements PrependEx
...
         'twig',
         'mailer',
+        'reports'
+    );
+
+    private $emails = array(
+        'order_comment',
+        'order_confirmation',
+        'customer_welcome',
     );
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/SyliusCoreExtension.php]: Please remove, wrong rebase.

**Comment by pjedrzejewski** [comment]
Thanks guys, nice work! :+1:

**Comment by a2xchip** [comment]
:+1:

**Pull request merged**